week1/search.py

```diff
@@ -109,13 +112,73 @@ def query():
 def create_query(user_query, filters, sort="_score", sortDir="desc"):
     print("Query: {} Filters: {} Sort: {}".format(user_query, filters, sort))
     query_obj = {
-        'size': 10,
-        "query": {
-            "match_all": {} # Replace me with a query that both searches and filters
+        "size": 13,
+        "query" : {
+            "function_score" : {
+                        "query": { "bool" : {
+                            "must" : {
+                                "query_string" : {
+                                    "fields" : ["name^100", "shortDescription^50", "longDescription^10", "department"],
+                                    "query" : user_query,
+                                    "phrase_slop": 3
+                                }
+                            },
+                            "filter": filters
+                            }
+                        },
+                        "boost_mode": "multiply",
+                        "score_mode": "avg",
+                        "functions": [
+                            {"field_value_factor": {
+                            "field": "salesRankShortTerm",
+                            "missing": 100000000,
+                            "modifier": "reciprocal"
+                            }},
+                            {"field_value_factor": {
+                            "field": "salesRankMediumTerm",
+                            "missing": 100000000,
+                            "modifier": "reciprocal"
+                            }},
+                            {"field_value_factor": {
+                            "field": "salesRankLongTerm",
+                            "missing": 100000000,
+                            "modifier": "reciprocal"
+                            }}
+                        ]
+            }
         },
         "aggs": {
-            #### Step 4.b.i: create the appropriate query and aggregations here
-
-        }
+            "regularPrice": {
+                "range": {
+                    "field": "regularPrice",
+                    "ranges": [
+                        {"to": 15.0},
+                        {"from": 15.0, "to": 50.0},
+                        {"from": 50.0, "to": 100.0},
+                        {"from": 100.0}
+                    ]
+                }
+             },
+            "department" : {
+                "terms": {
+                    "field": "department.keyword",
+                    "size": 10,
```

Comment:
I didn't limit the departments, because I wanted to show all of them, and there is a short list of them.

I did another aggregation for "brands" where I did use the "size" limit to retrieve only the brands with most items in them. I had to touch some the UI to see the brand list/filtering but the query part looks very much alike as your department aggregation query :) 

```
            "manufacturer": {
                "terms":{
                    "field": "manufacturer.keyword",
                    "size": 10,
                    "min_doc_count": 0,
                    "exclude": "UNKNOWN|Unknown"
                }
            },
```